[스프린트 2주차] singco(박재현) 미션 제출합니다.

SingTheCode opened the pull request:
스프린트 2주차에 대한 클래스 다이어그램과 요구사항 목록은 "요구사항.md" 에 들어가 있습니다.
감사합니다.

SingTheCode committed:
Added Untitled Diagram.drawio

SingTheCode committed:
feat: Log, Commander 클래스 간 의존성 주입

SingTheCode committed:
feat: 명령어와 number list를 분리하는 메서드 추가

SingTheCode committed:
docs: 요구사항 및 클래스 다이어그램 정리

SingTheCode committed:
feat: 명령어 따라 각각에 맞는 기능 실행 메서드 추가
- 명령어가 'count'일 때 count() 실행
- 명령어가 'add'일 때 add() 실행

SingTheCode committed:
Update Untitled Diagram.drawio

SingTheCode committed:
feat: custom Reduce 구현

SingTheCode committed:
docs: 요구사항 구현 현황을 최신화한다
- reduce 함수를 직접 만든다

SingTheCode committed:
feat: 커맨드 명령어에 따라 count() 메서드나 add() 메서드를 실행한다
- count(): 연달아 오는 값들의 개수를 센다
- add(): 연달아 오는 값들의 합을 구한다

SingTheCode committed:
chore: npm 초기 설정한다

SingTheCode committed:
chore: pacakge-lock.json을 생성한다

SingTheCode committed:
feat: 내부 메서드를 실행한다
- seperateNumbersCommands(), selectCommands() 를 실행한다

SingTheCode committed:
fix: 결과 값이 아무 곳에도 담기지 않는 문제를 해결한다

SingTheCode committed:
fix: 입력받은 숫자 배열이 string type으로 들어가는 문제를 해결한다
- numbers 배열에 Number 타입으로 넣는다

SingTheCode committed:
refactor: commander 클래스의 내부 명령어들을 실행하고 result를 Log 클래스로 넘겨줌으로써 클래스의 역할을 구분한다

SingTheCode committed:
feat: Commander 클래스의 처리된 결과 값을 string type으로 받아 출력한다

SingTheCode committed:
docs: 요구사항 구현 현황을 최신화한다
- 구현 마무리

SingTheCode committed:
Merge branch 'Park-Jaehyun' of https://github.com/SingTheCode/web-with-js into Park-Jaehyun

SingTheCode committed:
docs: 클래스 다이어그램을 최신화한다

gxxrxn commented on sprint-2/Untitled Diagram.drawio:
```diff
@@ -0,0 +1,60 @@
+      </root>
+    </mxGraphModel>
+  </diagram>
+</mxfile>
```
오.. 이건 뭐죠...
혹시 불필요한 파일이라면 git ignore 처리해주세요!!

SingTheCode replied:
draw.io로 클래스 다이어그램을 작성한 파일인데 저렇게 보이는 것 같습니다!ㅎㅎ

gxxrxn commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,21 @@
-Array.prototype.customReduce = function() {
-  return 0;
+Array.prototype.customReduce = function(callback, initialValue = 0) {
+  let array = Object(this);
```
this도 변하는 경우가 있나요...? const로 선언하면 좋을 것 같아요!

SingTheCode replied:
callback() 메서드에서 array도 optional parameter라 const array 선언해주면 const 에러가 뜨더라구요!! 아래에서 말씀하신 것 처럼 default parameter를 빼고 const로 바꿔보겠습니다!

gxxrxn commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,21 @@
+  let accumulator = array[0];
+  let index = 1;
+
+
```
공백도.. 컨벤션이다..!!

SingTheCode replied:
const와 let 사이에 공백을 주라는 말씀이신가요??

gxxrxn replied:
공백이 두줄이라 말씀드렸어요!!
불필요한 공백은 제거하면 좋을 것 같아서요!!

gxxrxn commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,21 @@
+
+  while(index < length) {
+    let currentValue = array[index];
+    accumulator = callback(accumulator, currentValue, currentIndex = index, array = array);
```
array는 원본 배열이니까 변하면 안될 것 같아요!
`accumulator = callback(accumulator, currentValue, currentIndex = index, array);`
이렇게 처리해도 되지 않을까요..?

SingTheCode replied:
오!! 규란님 말씀이 맞는 것 같아요!! currentIndex도 굳이 default parameter로 안해도 될 것 같습니다! 처리해보겠습니다!!

gxxrxn commented on sprint-2/domain/commander.js:
```diff
@@ -1,5 +1,51 @@
+    this.command = arr[0];
+    for(let i = 1; i < arr.length; i++) {
+      this.numbers.push(Number(arr[i]));
+    }
```
미션 요구사항 중 count 명령 값으로는 문자도 사용 가능하다고 되어있는데, Number로 모든 값을 바꿔버리면 안될 것 같아요!

SingTheCode replied:
요구사항을 체크하면서 했는데도 빼먹은 조건이 있었네요... 반성해야될 것 같습니다. fix 하도록 하겠습니다!

gxxrxn commented on sprint-2/domain/commander.js:
```diff
@@ -1,5 +1,51 @@
+
+  add() {
+    const addNumbers = this.numbers.customReduce((accumulator, currentValue) => accumulator + currentValue);
+    this.result = `add: ${addNumbers}`;
```
commander가 result를 이런 형식으로 가지는 것 보다 명령과 결과를 가지고 있다가 리턴해주고 Log를 통해 출력하는 것이 더 좋지 않을까요..?
나중에 출력 형식을 다르게 하고 싶을 때 commander를 바꾸는 것 보다는 Log를 수정하는 것이 더 바람직한 것 같아요!

SingTheCode replied:
확실히 string을 완성시켜주는 건 Log 클래스에서 하는게 나을 것 같네요! 감사합니다!

gxxrxn commented on sprint-2/요구사항.md:
```diff
@@ -0,0 +1,39 @@
+
+명령어의 종류에 따라 다른 기능을 하는 Commander 클래스 생성
+
+![img.png](img.png)
```
리드미에 사진이 깨져보여요!
링크 수정하면 좋을 것 같아요!

SingTheCode replied:
확인해보고 수정하겠습니다!

SingTheCode replied:
경로는 맞는데 merge되기 전 시점에는 해당 이미지 파일이 해당 경로에 없는걸로 인식이 되서 안보이시는 것 같아요! 제 fork repository에서의 요구사항.md에서는 이미지가 잘 보입니다! merge되고 한번 더 확인해주시면 감사하겠습니다!

gxxrxn commented on sprint-2/main.js:
```diff
@@ -1,4 +1,13 @@
+  const inputValue = 'count 1 2 3 4 5'
+
+  const commander = new Commander(inputValue);
+  commander.init();
```
init을 밖에서 해주는 이유가 있나요..?!?
생성자에서 해도 좋을 것 같아요!

SingTheCode replied:
Commander 내부 메서드 실행을 클래스 내부에서 해결하고 싶었는데 어디다 선언해야되는지를 몰라서 밖으로 뺐습니다. constructor 내부에서 메서드를 실행하는게 일반적인 방법일까요??

gxxrxn reviewed (commented):
draw.io로 클래스다이어그램까지 그리시다니..! 새로운 시도 멋지십니당..👍🏻
count 명령은 문자도 처리가 가능해야하는데, 현재 코드로는 오류가 발생할 것 같습니다!
수정해보시면 좋을 것 같아요!
수고하셨습니다!

SingTheCode reviewed (commented):
지금보니 요구사항.md에 적은 요구사항 목록에서 count()에 연달아오는 경우에 문자가 포함될 수 있다는 부분을 체크하지 못했네요...! 수정하겠습니다! Log에서 출력형식을 정해주는 방향으로 진행하면 좋겠다는 얘기가 특히 기억에 남습니다.
감사합니다!

SingTheCode committed:
fix: array(optional) parameter이 변경될 수 있는 문제를 해결한다
- default parameter로 처리하지 않고 array를 계속 parameter로 입력받게 한다
- let array를 const로 변경한다

SingTheCode committed:
fix: count 시 문자를 입력받으면 Number로 파싱할 때 생기는 에러를 해결한다.

SingTheCode committed:
refactor: 출력 형식을 log에서 담당할 수 있게끔 commander, log의 역할을 분리한다
- Commander에서 넘겨주는 result값을 출력 형식이 아닌 명령어와 결과값이 들어가있는 배열로 변경한다
- Log에서는 result 배열을 받아 출력 형식을 만든다

SingTheCode committed:
refactor: Commander 내부에서 실행되어야 할 메서드를 인스턴스 생성 시 처리하도록 하여 내부 동작을 숨긴다
- separateNumbersCommands(), selectCommands() 메서드를 constuctor 내부에 넣어 인스턴스 생성 시 실행되게끔 한다

SingTheCode committed:
refactor: 공백이 두 줄인 부분을 한 줄로 변경한다

2yunseong commented on sprint-2/domain/commander.js:
```diff
@@ -1,5 +1,48 @@
+    }
+  }
+
+  count() {
+    const countNumbers = this.numbers.customReduce((accumulator, currentValue, currentIndex) => Number(accumulator) + 1, 0);
+    this.result = ['count', countNumbers];
+  }
+
+  add() {
+    const addNumbers = this.numbers.customReduce((accumulator, currentValue) => Number(accumulator) + Number(currentValue));
+    this.result = ['add', addNumbers];
+  }
```
해당 콜백함수 내에서 customReduce를 사용하는 것보다는, customReduce 의 인자로 전달해야되지 않을까 싶습니다!
https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#%EC%98%88%EC%A0%9C

SingTheCode replied:
callback 함수로 count 함수를 사용해야된다는 얘기신가요?

SingTheCode replied:
저는 count는 Commander 인스턴스의 메서드이고 customReduce를 활용해서 count 메서드를 구현한다고 생각했습니다! 윤성님 말씀대로 코드를 짜보는 것도 좋을 것 같네요ㅎㅎ

2yunseong commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,20 @@
-Array.prototype.customReduce = function() {
-  return 0;
+Array.prototype.customReduce = function(callback, initialValue = 0) {
+  const array = Object(this);
+  const length = array.length;
+  let accumulator = array[0];
+  let index = 1;
+
+  if (initialValue !== 0) {
+    accumulator = initialValue;
+    index = 0;
+  }
+
+  while(index < length) {
+    let currentValue = array[index];
+    accumulator = callback(accumulator, currentValue, index, array);
+    index++;
+  }
+  return accumulator;
```
customReduce는 initialValue를 optional로 받을 수 있습니다.
따라서, initialValue를 인자로 전달 받을 때와, 받지 않을 때의 로직의 구분이 있어야 한다 생각합니다.
그 이유는 다음과 같습니다.

- initalValue를 제공한 경우
    - accumulator 는 initialValue와 같다.
    - currentValue 는 배열의 첫 번째 값과 같다.
- initalValue를 제공하지 않은 경우
    - accumulator 는 배열의 첫 번째 값과 같다.
    - currentValue 는 배열의 두 번째 값과 같다.
- 배열이 비어있는데 initialValue를 제공한 경우
    - TypeError 발생.
- initialValue는 주어졌으나 배열이 비어있는 경우 & 배열의 요소가 1개 지만, initialValue를 제공하지 않은 경우
    - 단독 값을 callback 호출없이 반환

참고 1의 **설명** 부분을 제가 정리해보았는데, 같이 고민해보면 좋을 것 같아요. 틀린 점이 있다면 알려주세요.


----------
참고 1 : https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#%EC%84%A4%EB%AA%85

SingTheCode replied:
좋은 설명 감사합니다. reduce의 4가지 경우에 대해 자세히 알게된 것 같아요ㅎㅎ

initialValue를 default parameter로 설정하면서 initialValue가 optional로 들어올 수 있습니다. 

1, 2, 3, 4 케이스 중 1, 2 케이스는 initialValue !== 0 의 조건으로 처리했고, 3번 케이스는 현재는 SyntaxError로 나타나는데 Error 처리를 따로 해준다면 TypeError도 현재 코드에서 처리할 수 있을 듯 합니다.

2yunseong commented on sprint-2/main.js:
```diff
@@ -1,4 +1,12 @@
-function main() {}
+
+function main() {
+  const inputValue = 'count 1 2 3 4 5'
```
저는 세미콜론을 사용하면, 일관성을 위해 모든 코드에 사용해주는 것이 바람직하다고 생각합니다!
하지만 자바스크립트와 파이어폭스를 만든 브랜든 아이크는 세미콜론 사용을 반대하고 있다고 합니다.ㅋ
물론 안써주어도 자바스크립트 엔진이 세미콜론 자동 삽입 기능(automatic semicolon insertion)이 암묵적으로 삽입해준다고 하네요.
하지만 코드의 예측 가능성을 위해, 세미콜론을 모든 코드에 사용하는 것이 좋다고 저는 생각합니다! (자체 종결성을 가지는 if, for, 함수는 제외)
예측 불가능한 상황은 다음의 예제가 있습니다. 물론 이런 상황을 만들기는 어렵지만요.
```javascript
function foo() {
    return
        {}
// asi 의 동작 결과 : return; {};
// 개발자의 예측 : return {};
}
```
(다른 예시는 아래 참고 2 링크를 참고 해주시기 바랍니다. 칼럼 읽는 느낌으로 읽으시면 좋을 것 같아요.)

~~저도 세미콜론을 사용하는데, 자주 빼먹어서 vscode의 prettier가 채워주긴 합니다~~

----
참고 1 : 모던 자바스크립트 Deep Dive 5.5 세미콜론과 세미콜론 자동 삽입 기능
참고 2 : https://monaq.github.io/2018/01/28/semicolon/

SingTheCode replied:
요즘 알고리즘으로 파이썬을 하다보니깐 파이썬에 세미콜론 쓰고 js에서 세미콜론을 빼먹는 경우가 계속 생기더라구요ㅎㅎ( {} 중가로도...ㅎㅎ)

그것하고는 별개로 세미콜론을 왜 안써도 에러가 안뜨지라는 것에 대해 의문을 가지로 있었는데 코드의 예측 가능성과 js엔진의 세미콜론 자동 삽입 기능에 대해 설명해주셔서 해결이 됐습니다! 좋은 설명 감사합니다!!

2yunseong reviewed (commented):
고생하셨습니다! :D
